crates/sui-node/src/lib.rs

```diff
@@ -200,68 +202,87 @@ impl SuiNode {
         Ok(node_one_cell.get().await)
     }
 
-    fn start_jwk_updater(epoch_store: Arc<AuthorityPerEpochStore>) {
+    fn start_jwk_updater(
+        epoch_store: Arc<AuthorityPerEpochStore>,
+        consensus_adapter: Arc<ConsensusAdapter>,
+    ) {
         let epoch = epoch_store.epoch();
-        tokio::task::spawn(
-            async move {
-                info!("Starting JWK updater task");
-                loop {
-                    let epoch_store_ = epoch_store.clone();
-                    let supported_providers = epoch_store
-                        .protocol_config()
-                        .zklogin_supported_providers()
-                        .iter()
-                        .map(|s| OIDCProvider::from_str(s).expect("Invalid provider string"))
-                        .collect::<Vec<_>>();
-                    let fetch_and_sleep = async move {
-                        // Update the JWK value in the authority server
-                        info!("fetching new JWKs");
-                        match Self::fetch_jwks(&supported_providers).await {
-                            Err(e) => {
-                                warn!("Error when fetching JWK {:?}", e);
-                                // Retry in 30 seconds
-                                tokio::time::sleep(Duration::from_secs(30)).await;
-                            }
-                            Ok(keys) => {
-                                for (jwk_id, jwk) in keys {
-                                    epoch_store_.insert_oauth_jwk(&jwk_id, &jwk);
+        tokio::task::spawn(async move {
+            let supported_providers = epoch_store
+                .protocol_config()
+                .zklogin_supported_providers()
+                .iter()
+                .map(|s| OIDCProvider::from_str(s).expect("Invalid provider string"))
+                .collect::<Vec<_>>();
+
+            info!(
+                "Starting JWK updater task with supported providers: {:?}",
+                supported_providers
+            );
+
+            let mut join_set = tokio::task::JoinSet::new();
+            for p in supported_providers.into_iter() {
+                let epoch_store = epoch_store.clone();
+                let consensus_adapter = consensus_adapter.clone();
+                join_set.spawn(
+                    async move {
+                        // note: restart-safe de-duplication happens after consensus, this is
+                        // just best-effort to reduce unneeded submissions.
+                        let mut seen = HashSet::new();
```

Comment:
Should we at least load currently active JWKs? Otherwise it seems every time node restart it will re-send all jwks to consensus?

Reply:
That shouldn't be very many messages, but it will be easy enough to handle this.